Refactor Guardian.DB.Sweeper

doomspork opened the pull request:
I'd like to slowly start improving the code quality and test coverage across the Ueberauth org projects. Since we've been focusing on Guardian.DB recently I thought I'd start by proposing some changes here. For this PR I looked to tackle the Sweeper functionality first:

1. We currently have `Sweeper` and `SweeperServer` but their roles are not clearly separated (`Sweeper` manages the `SweeperServer`'s state for instance). I combined these together as there's no real benefit to keeping them separate and this cleans up some of their code.
2. Updated the tests to use the GenServer directly and test the 2 major functions it provides: `purge/0` and `reset_timer/0`.
3. Moved the Sweeper from `Guardian.DB.Token.SweeperServer` to `Guardian.DB.Sweeper`, I'm not entirely sure why this was ever nested under Token but it seems rather unnecessary. __This is a breaking change__.
4. Stop accessing configuration directly from the library code. It's generally a better practice for the application to have its own configuration which it passes in rather than libraries doing that themselves.

I may look at adding some addition test coverage for each of the GenServer callbacks just to cover our bases.

This PR is a place to start this conversation, hence the draft. :tada:

doomspork committed:
Refactor Guardian.DB.Sweeper

Moved the Sweeper functionality out from inside the Token
namespace/folder. Also simplified the code by combining the two separate
but unnecessary modules. Updated and added test coverage.

doomspork commented on test/guardian/sweeper_test.exs:
```diff
@@ -1,31 +1,46 @@
+      [interval: ^interval, timer: timer] = :sys.get_state(pid)
+
+      Sweeper.reset_timer()
+
```
I'll add a `GenServer.stop(pid)` here :+1:

alexfilatov commented on lib/guardian/db/sweeper.ex:
```diff
@@ -0,0 +1,70 @@
+  @sixty_minutes 60 * 60 * 1000
+
+  def start_link(opts) do
+    interval = Keyword.get(opts, :interval, @sixty_minutes)
```
looks like it's `: sweep_interval` in the config, @doomspork 

```
config :guardian, Guardian.DB,
  repo: MyApp.Repo, # Add your repository module
  schema_name: "guardian_tokens", # default
  token_types: ["refresh_token"], # store all token types if not set
  sweep_interval: 60 # default: 60 minutes
```

alexfilatov replied:
oh I got it now, so it's not configurable anymore?

doomspork replied:
It is configurable @alexfilatov but as I mentioned in this point:

> 4. Stop accessing configuration directly from the library code. It's generally a better practice for the application to have its own configuration which it passes in rather than libraries doing that themselves.

I think we should be moving away from our libraries accessing application configuration themselves directly. Applications should be able to pass in their configurations. 

I'm open to _not_ doing this, I think it's just a better design practice. 


```elixir
children = [
  ...
  {Guardian.DB.Sweeper, interval: configured_sweep_interval()}
  ... 
]

```

yordis replied:
I love this part for sure

yordis commented:
> We currently have Sweeper and SweeperServer but their roles are not clearly separated (Sweeper manages the SweeperServer's state for instance). I combined these together as there's no real benefit to keeping them separate and this cleans up some of their code.

I disagree with the promise of no real benefits.

SweeperServer is impure, it deals with Erlang message passing and gen-server stuff. While Sweeper is pure since it is just the implementation of the Sweeper. Functional Core with Imperative Interface.

That being said, you are the boss. 

I just would try to be clear on what is wrong as of today based on the objective issues since it wasn't been a problem so far before we start removing things just because it is deemed unimportant based on your skills set or your perspective, which it isn't hard to understand where you are coming from for sure. I totally understand your reasoning.

yordis commented on test/guardian/sweeper_test.exs:
```diff
@@ -1,31 +1,46 @@
+
+      Sweeper.reset_timer()
+
+      [interval: _interval, timer: new_timer] = :sys.get_state(pid)
```
This is what I am talking about with impure vs pure. The test case is around the impurity of the genserver where we can test the logic without having to reach for such a layer and simply test the return value.

At some point, testing the Server doesn't even matter since we are just testing Erlang GenServer abstraction, and the value of it is diminished since we tested the behavior using the Sweeper, and SweeperServer was just the message passing from Erlang 

🤷🏻

doomspork replied:
This is intentional. I _want_ to test that the GenServer works like we expect. As I mentioned in my PR description, I can add tests to handle the callbacks which would address your concern.

yordis commented:
Overall, don't mind the changes, I can maintaining them either way, if you feel strong about it, you have my 1+

doomspork commented:
> That being said, you are the boss.

It's not that at all. I'm open to ideas and suggestions.

>  just would try to be clear on what is wrong as of today based on the objective issues since it wasn't been a problem so far before we start removing things just because it is deemed unimportant based on your skills set or your perspective, which it isn't hard to understand where you are coming from for sure. I totally understand your reasoning.

I'm not sure "wasn't been a problem so far" is ever good reason to avoid changes, would we ever improve code with that mentality? If your concern is having to test the GenServers I can update the test to use the GenServer callbacks directly and avoid that. I actually _want_ to test that this stuff works accordingly so these tests are intentional but I'm happy to remove them.

Simplifying the code to be easier to understand makes things better for everyone, _especially_ us maintainers. A code base that's difficult to grok keeps contributors away.

I'm okay to not incorporating this change but I think it's misguided to assume that just because we can have "pure" functions to test that there isn't a better solution.  As-is responsibility is spread throughout modules making understanding the code base far more complicated/

doomspork committed:
Expand on test coverage

yordis commented:
Works for me either way to me, I leave this one to your judgment, I trust your judgments.

doomspork commented:
Once #129 is merged I will update this further :+1:

doomspork committed:
Call GenServer callbacks directly

doomspork commented:
@yordis I updated the tests and made use of Mox for 2 reasons:

1. Updated the tests to use the GenServer callbacks directly. This should address you concern with impure functions.
2. Using Mox we can test the functionality works without having to actually interact with the database.

doomspork commented on lib/guardian/db/sweeper.ex:
```diff
@@ -0,0 +1,70 @@
+  scheduled work.
+  """
+  def purge do
+    GenServer.cast(__MODULE__, :sweep)
```
These were also changed to `cast` from `call` since we don't need or care about a reply.

doomspork commented on lib/guardian/db/sweeper.ex:
```diff
@@ -0,0 +1,70 @@
+defmodule Guardian.DB.Sweeper do
```
For backwards compatibility we could add a `Guardian.DB.Token.SweeperServer` alias but I'd like to encourage folks to move away from that module.

yordis replied:
@doomspork wouldn't be better to use `SweeperServer` following some `GenServer` lingos?

It is a stateful thing (hint the server part of it) that you send messages to it 🤔 💭

doomspork replied:
I don't often seen things called "Server" in Elixir/Erlang space, do you? Looking at a few of my applications there's nothing in the application tree with that suffix :thinking:

yordis replied:
It depends if you have the `Server` being the message passing, while the thing without `Server` (as it was before), was the implementation of the thing.

Dave has an amazing course about it, https://codestool.coding-gnome.com/courses/elixir-for-programmers

doomspork replied:
I appreciate Dave's pattern but as much as he would like to see this catch on I can't say I see this having gained any adoption. I prefer to keep code simple and easy to grok for the human reader over trying to be "academically right".

I'd like to leave this.
